src/iterators/chunked_read_ahead_iterator.rs

```diff
@@ -0,0 +1,368 @@
+    /// # Panics
+    ///
+    /// - panics if the underlying iterator panics
+    /// - panics if sending final chunk after an internal panic fails
+    /// - panics if sending the None kill signal to the recieiver after an internal panic fails
+    /// - panics if the spawned thread fails to spawn
+    pub fn new<I>(mut inner: I, chunk_size: usize, num_chunk_buffer_size: usize) -> Self
+    where
+        I: Iterator<Item = T> + Send + 'static,
+    {
+        assert_ne!(chunk_size, 0, "Chunk size cannot be zero!");
+        assert_ne!(num_chunk_buffer_size, 0, "Number of buffered chunks cannot be zero!");
+
+        // Create a channel over which we can send our chunks of ``T``
+        let (sender, receiver) = sync_channel(num_chunk_buffer_size);
+
+        // Create our spawned thread, holding on to the resulting handle for downstream error
+        // management.
+        thread::Builder::new()
+            .name("chunked_read_ahead_thread".to_owned())
+            .spawn(move || {
+                'chunk_loop: loop {
+                    let mut chunk = Vec::with_capacity(chunk_size);
+                    for _ in 0..chunk_size {
+                        match catch_unwind(AssertUnwindSafe(|| inner.by_ref().next())) {
+                            Ok(Some(val)) => chunk.push(val),
+                            Ok(None) => break,
+                            Err(e) => {
+                                sender
+                                    .send(Ok(chunk))
+                                    .expect("Error sending final chunk before internal panic");
+                                sender.send(Err(e)).expect("Error sending internal panic result");
+                                break 'chunk_loop;
+                            }
+                        }
+                    }
+                    if chunk.is_empty() {
+                        break;
+                    }
+                    let send_result = sender.send(Ok(chunk));
+                    if send_result.is_err() {
+                        // TODO - Logging?
+                        break;
+                    }
+                }
+            })
+            .expect("failed to spawn chunked read ahead thread");
+
+        // Store the necessary objects on ``Self``
+        Self { current_chunk: Vec::new().into_iter(), receiver: Some(receiver) }
+    }
+}
+
+impl<T> Drop for ChunkedReadAheadIterator<T>
+where
+    T: Send + 'static,
+{
+    fn drop(&mut self) {
+        // Make sure our reciever is dropped so our spawned thread shuts down.
+        self.receiver = None;
+    }
+}
+
+impl<T> Iterator for ChunkedReadAheadIterator<T>
+where
+    T: Send + 'static,
+{
+    type Item = T;
+
+    fn next(&mut self) -> Option<T> {
+        // Check if our current chunk has anything left in it
+        // If so, just return that result
+        // If not, see documentation on else block
+        if let Some(result) = self.current_chunk.next() {
+            Some(result)
```

Comment:
It always feels weird to me to unwrap and then re-wrap a Some.  Would it be better to do:


```rust
let next_option = self.current_chunk.next();
if next_option.is_some() {
    next_option
}
else {
    ....
}
```